File: LeetCode/LinkedList_Cycle_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool detectCycle(ListNode *head) {
        if(!head) return false;
        ListNode *slow = head;
        ListNode *fast = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
            if(slow == fast)
                break;
        }
        if(!fast || !fast->next)
            return false;
        else{
            slow = head;
            while(slow != fast){
                slow = slow->next;
                fast = fast->next;
            }
            return true;
        }
    }
    
    bool hasCycle(ListNode *head) {
        return detectCycle(head);
    }
};




    

File: LeetCode/linked_list_cycle_2.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }3, 2, 0, -4
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null) return null;
        ListNode slowIter = head, fastIter = head;
        
        while(true){
            if(fastIter != null && fastIter.next != null){
                fastIter = fastIter.next.next;
                slowIter = slowIter.next;
            }else{
                break;
            }
            
            if(slowIter == fastIter) break;
        }
        
        if(fastIter == null || fastIter.next == null) return null;
        
        slowIter = head;
        while(slowIter != fastIter){
            fastIter = fastIter.next;
            slowIter = slowIter.next;
        }
        return fastIter;
    }
}